packet03-redirecting/README.org

```diff
@@ -5,49 +5,277 @@
   - [[#assignment-3-extend-to-a-bidirectional-router][Assignment 3: Extend to a bidirectional router]]
   - [[#assignment-4-use-the-bpf-helper-for-routing][Assignment 4: Use the BPF helper for routing]]
 
+* What you will learn in this lesson
+
+** Sending packets back to the interface they came from
+
+The =XDP_TX= return value can be used to send the packet back from the same
+interface it came from.  This functionality can be used to implement load
+balancers, to send simple ICMP replies, etc.  We will use this functionality in
+the Assignment 1 to implement a simple ICMP echo server.
+
+Note that in all our assignments below we are using the =skb= (i.e.,
+=xdpgeneric=) mode to load programs.  We are doing this because =veth= devices
+won't deliver redirected/retransmitted XDP frames unless there is an XDP
+program attached to the receiving side of the target =veth= interface. Physical
+hardware will likely behave the same. XDP maintainers are currently working on
+fixing this behaviour upstream.
+
+** Redirecting packets to other interfaces
+
+Besides the ability to transmit packets back from the same interface, there is
+an option to forward packets to egress ports of other interfaces (if the
+corresponding driver supports this feature). This can be done using the
+=bpf_redirect= or =bpf_redirect_map= helpers. These helpers will return the
+=XDP_REDIRECT= value and this is what should the program return. The
+=bpf_redirect= helper actually shouldn't be used in production as it is slow
+and can't be configured from user space.  However, we will use it in the
+Assignment 2 for the sake of simplicity. To use the =bpf_redirect_map= helper
+we need to setup a special map of type =BPF_MAP_TYPE_DEVMAP= which maps virtual
+ports to actual network devices. See [[https://lwn.net/Articles/728146][this
+patch series]] which implemented the XDP redirect functionality. An example of
+usage can be found in the
+[[https://github.com/torvalds/linux/blob/master/samples/bpf/xdp_redirect_map_kern.c][xdp_redirect_map_kern.c]]
+file and the corresponding loader.
+
+** Building a router using a kernel helper
+
+As we will see in the Assignments 2 and 3, using the raw redirect functionality
+could be quite challenging, as one needs to support a set of data structures
+and a custom kernel-to-user space protocol to exchange data.  Fortunately, for
+XDP and TC programs we can access the kernel routing table using the
+=bpf_fib_lookup()= helper and use this information to redirect packets.  The
+Assignment 4 shows how to use this helper to write a program for full
+redirection.
+
+One thing to note when using the =bpf_fib_lookup()= function is that it will
+return an error code if forwarding is disabled, so enable forwarding before
+running code from the Assignment 4.
+#+begin_src sh
+$ sudo sysctl net.ipv6.conf.all.forwarding=1
+net.ipv6.conf.all.forwarding = 1
+$ sudo sysctl net.ipv4.ip_forward=1
+net.ipv4.ip_forward = 1
+#+end_src
```

Comment:
I think we can just enable forwarding per-interface in the venv setup code in testenv.sh, can't we? I'd keep the note about forwarding needing to be enabled, but would rather not instruct users to make global changes to their machine just to run the tutorial...

Reply:
Ok, per-interface forwarding is better. But I would rather do it as a part of solution. It is fun that we can redirect packets in the Assignments 2 and 3 without enabling forwarding on interfaces. What do you think?

Reply:
Right, good point about demonstrating that we can redirect without forwarding being turned on. But in that case we should probably point that out somewhere ;)

Reply:
Ah, I think that I mentioned that, but it was lost when I addressed your first series of comments